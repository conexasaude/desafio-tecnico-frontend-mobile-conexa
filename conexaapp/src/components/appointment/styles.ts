import styled from 'styled-components/native';

export const Container = styled.TouchableOpacity`
  flex: 1;
  background-color: ${({theme}) => theme.COLORS.WHITE};
  border-radius: 6px;
  margin-bottom: 10px;
  padding: 10px;
`;

export const PatientContainer = styled.View`
  gap: 5px;
`;

export const DateHour = styled.Text`
  align-self: flex-end;
  font-weight: 500;
  font-size: ${({theme}) => theme.FONT_SIZE.BODY.XS}px;
`;

export const Patient = styled.Text`
  font-weight: 500;
  font-size: ${({theme}) => theme.FONT_SIZE.BODY.M}px;
`;

export const Description = styled.Text`
  font-size: ${({theme}) => theme.FONT_SIZE.BODY.S}px;
`;

export const DoctorContainer = styled.View`
  border-top-width: 1px;
  border-color: ${({theme}) => theme.COLORS.GRAY_300};
  margin-top: 5px;
  padding-top: 5px;
`;

export const DoctorName = styled.Text`
  font-size: ${({theme}) => theme.FONT_SIZE.BODY.S}px;
`;

export const DoctorEmail = styled.Text`
  font-size: ${({theme}) => theme.FONT_SIZE.BODY.S}px;
`;

export const TextBold = styled.Text`
  font-weight: bold;
`;
