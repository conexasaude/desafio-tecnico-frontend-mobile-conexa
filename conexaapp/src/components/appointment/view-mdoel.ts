import {useNavigation} from '@react-navigation/native';
import {AppNavigatorRoutesProps} from '@routes/app.routes';

interface handlePressProps {
  id: number;
}

export interface AppointmentViewModel {
  handlePress: (value: handlePressProps) => Promise<void>;
}

function useAppointmentViewModel(): AppointmentViewModel {
  const {navigate} = useNavigation<AppNavigatorRoutesProps>();

  async function handlePress({id}: handlePressProps) {
    navigate('appointment', {id});
  }

  return {
    handlePress,
  };
}

export {useAppointmentViewModel};
