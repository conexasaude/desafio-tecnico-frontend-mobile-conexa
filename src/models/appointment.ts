export interface Appointment {
  id: number;
  medico: Doctor;
  paciente: string;
  dataConsulta: string;
  observacao: string;
}

interface Doctor {
  id: number;
  nome: string;
  email: null;
}
