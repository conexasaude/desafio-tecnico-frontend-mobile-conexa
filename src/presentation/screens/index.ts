export * from './appointments-screen';
export * from './create-appointment-screen';
export * from './login-screen';
