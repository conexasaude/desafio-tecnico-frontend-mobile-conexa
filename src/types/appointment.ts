export type Appointment = {
	dataConsulta: string
    id: number
    medico: {
        id: number,
        nome: string,
        email: null
    }
    observacao: string
    paciente: string
}

export type AppointmentForm = {
    idMedico: number
    paciente: string
	dataConsulta: Date
    observacao: string
}