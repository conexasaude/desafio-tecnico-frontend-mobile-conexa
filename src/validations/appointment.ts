import { z } from 'zod';

const appointmentSchema = z.object({
  patient: z.string({
    required_error: 'O campo nome do paciente é obrigatório',
  }),
  date: z.string({ required_error: 'O campo data é obrigatório' }),
  time: z.string({ required_error: 'O campo hora é obrigatório' }),
  observation: z.string().optional(),
});

export type FormScheduleAppointment = z.infer<typeof appointmentSchema>;

export default appointmentSchema;
