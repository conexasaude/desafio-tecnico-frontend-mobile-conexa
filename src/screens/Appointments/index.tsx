import React, { useCallback, useEffect } from 'react';
import { useReduxSelector } from '@/hooks/useReduxSelector';
import { useReduxDispatch } from '@/hooks/useReduxDispatch';
import { FlatList } from 'react-native';
import { Card } from '@/components/Card';
import { useTheme } from '@emotion/react';
import { useNavigation } from '@react-navigation/native';
import { NavPropsLogged } from '@/routes/logged';
import { fetchAppointments } from '@/store/thunks/appointment.thunk';
import { useToast } from 'react-native-toast-notifications';
import { Loader } from '@/components/Loader';
import { ErrorBoundary } from '@/components/ErrorBoundary';
import * as Styled from './styles';

export function AppointmentsScreen() {
  const { show } = useToast();
  const { effects } = useTheme();
  const { navigate } = useNavigation<NavPropsLogged>();
  const { appointments, loading, error } = useReduxSelector(
    state => state.appointment,
  );

  const dispatch = useReduxDispatch();

  const getAppointments = useCallback(() => {
    dispatch(fetchAppointments({ onError: show }));
  }, [dispatch, show]);

  useEffect(() => {
    getAppointments();
  }, [getAppointments]);

  if (loading) {
    return <Loader />;
  }

  if (!loading && error) {
    return <ErrorBoundary tryAgain={getAppointments} />;
  }

  return (
    <Styled.SafeArea>
      <Styled.Container>
        <Styled.Column>
          <Styled.Title>Painel de Consultas</Styled.Title>
          <Styled.Line />
        </Styled.Column>

        <Styled.Row>
          <Styled.DescriptionList>{`${appointments.length} consulta(s) agendadas`}</Styled.DescriptionList>
          <Styled.ScheduleButton
            title="Agendar Consulta"
            onPress={() => navigate('ScheduleAppointment')}
          />
        </Styled.Row>

        <FlatList
          data={appointments}
          keyExtractor={({ id }) => id.toString()}
          renderItem={({ item }) => (
            <Card
              testID="appointment-card"
              title={item.paciente}
              description={item.dataConsulta}
              button={{
                title: 'Ver Detalhes',
                onPress: () => navigate('DetailsAppointment', { id: item.id }),
              }}
            />
          )}
          contentContainerStyle={{ paddingHorizontal: effects.spacing.md }}
        />
      </Styled.Container>
    </Styled.SafeArea>
  );
}
