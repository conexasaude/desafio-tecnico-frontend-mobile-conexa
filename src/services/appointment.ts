import { Appointment, AppointmentForm } from "../types/appointment";
import api from "./api";

export async function getAppointments(): Promise<Appointment[]> {
	const response = await api.get('consultas', { params: { authRequired: true } })

	return response.data.data
}

export async function newAppointment(body: AppointmentForm): Promise<void> {
	const response = await api.post('consulta', {
		...body
	}, { params: { authRequired: true } })
	
	return response.data.data
}