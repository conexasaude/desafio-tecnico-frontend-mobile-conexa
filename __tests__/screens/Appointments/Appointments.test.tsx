import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/lib/functions/Testing';

import { AppointmentsScreen } from '@/screens/Appointments';
import { fireEvent } from '@testing-library/react-native';
import { AppointmentState } from '@/store/slices/appointment.slice';

const mockDispatch = jest.fn();
const mockNavigate = jest.fn();

const appointments = [
  {
    id: 16,
    medico: {
      id: 1,
      nome: 'Dr. Daniel Vieira',
      email: null,
    },
    paciente: 'Diego Senna',
    dataConsulta: '2020-04-23 08:30',
    observacao: 'Exemplo de consulta',
  },
  {
    id: 11,
    medico: {
      id: 1,
      nome: 'Dr. Daniel Vieira',
      email: null,
    },
    paciente: 'Rafael Braga',
    dataConsulta: '2020-04-23 08:30',
    observacao: 'Exemplo de consulta',
  },
];

jest.mock('@/hooks/useReduxDispatch', () => ({
  useReduxDispatch: () => mockDispatch,
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockNavigate }),
  useIsFocused: () => false,
}));

describe('Appointments', () => {
  const mockedAppointmentStore: AppointmentState = {
    loading: false,
    error: undefined,
    appointments,
  };

  test('should render correctly', () => {
    renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedAppointmentStore,
    });
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<AppointmentsScreen />, {
      appointment: mockedAppointmentStore,
    });
  });

  test('should render all cards', () => {
    const { getAllByTestId } = renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedAppointmentStore,
    });
    const appointmentsCard = getAllByTestId('appointment-card');
    expect(appointmentsCard.length).toEqual(2);
  });

  test('should call navigate to schedule appointment', () => {
    const { getByText } = renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedAppointmentStore,
    });
    const scheduleAppointmentButton = getByText('Agendar Consulta');
    fireEvent.press(scheduleAppointmentButton);
    expect(mockNavigate).toHaveBeenCalledWith('ScheduleAppointment');
  });

  test('should call navigate to details appointment', () => {
    const { getAllByTestId } = renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedAppointmentStore,
    });
    const appointmentsCard = getAllByTestId('appointment-card');

    appointmentsCard.forEach((card, index) => {
      fireEvent.press(card);
      expect(mockNavigate).toHaveBeenCalledWith('DetailsAppointment', {
        id: appointments[index].id,
      });
    });
  });

  test('should render Loader', () => {
    const mockedLoadingAppointmentStore: AppointmentState = {
      appointments: [],
      error: undefined,
      loading: true,
    };

    const { getByTestId } = renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedLoadingAppointmentStore,
    });

    expect(getByTestId('loader')).toBeDefined();
  });

  test('should render ErrorBoundary', () => {
    const mockedErrorAppointmentStore: AppointmentState = {
      appointments: [],
      error: 'Something went wrong',
      loading: false,
    };

    const { getByTestId } = renderWithThemeProvider(<AppointmentsScreen />, {
      appointment: mockedErrorAppointmentStore,
    });

    expect(getByTestId('error-boundary')).toBeDefined();
  });
});
