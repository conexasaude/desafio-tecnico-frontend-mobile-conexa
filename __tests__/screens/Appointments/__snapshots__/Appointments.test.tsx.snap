// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appointments should matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "marginTop": 40,
            "paddingHorizontal": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#2E50D4",
              "fontFamily": "Nunito-Bold",
              "fontSize": 24,
            },
            undefined,
          ]
        }
      >
        Painel de Consultas
      </Text>
      <View
        style={
          [
            {
              "backgroundColor": "#E5E5E5",
              "height": 1,
              "marginBottom": 40,
              "marginTop": 16,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      style={
        [
          {
            "alignItems": "flex-end",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 32,
            "paddingHorizontal": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#7F7F7F",
              "fontFamily": "Nunito-SemiBold",
            },
            undefined,
          ]
        }
      >
        2 consulta(s) agendadas
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "backgroundColor": "#2E50D4",
            "borderColor": "#2E50D4",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "height": 52,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 8,
            "width": "40%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Nunito-Bold",
                "fontSize": 14,
                "textAlign": "center",
              },
              undefined,
            ]
          }
          variant="contained"
        >
          Agendar Consulta
        </Text>
      </View>
    </View>
    <RCTScrollView
      contentContainerStyle={
        {
          "paddingHorizontal": 16,
        }
      }
      data={
        [
          {
            "dataConsulta": "2020-04-23 08:30",
            "id": 16,
            "medico": {
              "email": null,
              "id": 1,
              "nome": "Dr. Daniel Vieira",
            },
            "observacao": "Exemplo de consulta",
            "paciente": "Diego Senna",
          },
          {
            "dataConsulta": "2020-04-23 08:30",
            "id": 11,
            "medico": {
              "email": null,
              "id": 1,
              "nome": "Dr. Daniel Vieira",
            },
            "observacao": "Exemplo de consulta",
            "paciente": "Rafael Braga",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#fff",
                "borderRadius": 8,
                "elevation": 2,
                "marginBottom": 16,
                "opacity": 1,
                "padding": 12,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.25,
                "shadowRadius": 1.5,
                "width": "100%",
              }
            }
            testID="appointment-card"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "flex": 1,
                      "height": 52,
                      "justifyContent": "space-around",
                      "marginRight": 4,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#7F7F7F",
                        "fontFamily": "Nunito-SemiBold",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                >
                  Diego Senna
                </Text>
                <Text
                  numberOfLines={2}
                  style={
                    [
                      {
                        "color": "#999999",
                        "fontFamily": "Nunito-Regular",
                        "fontSize": 14,
                      },
                      undefined,
                    ]
                  }
                >
                  2020-04-23 08:30
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#2E50D4",
                    "borderRadius": 8,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "height": 48,
                    "justifyContent": "center",
                    "opacity": 1,
                    "padding": 8,
                    "width": "auto",
                  }
                }
                testID="card-button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#2E50D4",
                        "fontFamily": "Nunito-Bold",
                        "fontSize": 14,
                        "textAlign": "center",
                      },
                      undefined,
                    ]
                  }
                  variant="outlined"
                >
                  Ver Detalhes
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "#fff",
                "borderRadius": 8,
                "elevation": 2,
                "marginBottom": 16,
                "opacity": 1,
                "padding": 12,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.25,
                "shadowRadius": 1.5,
                "width": "100%",
              }
            }
            testID="appointment-card"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "flex": 1,
                      "height": 52,
                      "justifyContent": "space-around",
                      "marginRight": 4,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    [
                      {
                        "color": "#7F7F7F",
                        "fontFamily": "Nunito-SemiBold",
                        "fontSize": 16,
                      },
                      undefined,
                    ]
                  }
                >
                  Rafael Braga
                </Text>
                <Text
                  numberOfLines={2}
                  style={
                    [
                      {
                        "color": "#999999",
                        "fontFamily": "Nunito-Regular",
                        "fontSize": 14,
                      },
                      undefined,
                    ]
                  }
                >
                  2020-04-23 08:30
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "borderColor": "#2E50D4",
                    "borderRadius": 8,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "height": 48,
                    "justifyContent": "center",
                    "opacity": 1,
                    "padding": 8,
                    "width": "auto",
                  }
                }
                testID="card-button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#2E50D4",
                        "fontFamily": "Nunito-Bold",
                        "fontSize": 14,
                        "textAlign": "center",
                      },
                      undefined,
                    ]
                  }
                  variant="outlined"
                >
                  Ver Detalhes
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
